src/query.cpp

```diff
@@ -0,0 +1,266 @@
+{
+    QueryContainer query;
+    query.m_data->term_ids = std::move(term_ids);
+    return query;
+}
+
+auto QueryContainer::id() const noexcept -> std::optional<std::string> const&
+{
+    return m_data->id;
+}
+auto QueryContainer::string() const noexcept -> std::optional<std::string> const&
+{
+    return m_data->query_string;
+}
+auto QueryContainer::terms() const noexcept -> std::optional<std::vector<std::string>> const&
+{
+    return m_data->processed_terms;
+}
+
+auto QueryContainer::term_ids() const noexcept -> std::optional<std::vector<std::uint32_t>> const&
+{
+    return m_data->term_ids;
+}
+
+auto QueryContainer::threshold() const noexcept -> std::optional<float> const&
+{
+    return m_data->threshold;
+}
+
+auto QueryContainer::string(std::string raw_query) -> QueryContainer&
+{
+    m_data->query_string = std::move(raw_query);
+    return *this;
+}
+
+auto QueryContainer::parse(ParseFn parse_fn) -> QueryContainer&
+{
+    if (not m_data->query_string) {
+        throw std::domain_error("Cannot parse, query string not set");
+    }
+    auto parsed_terms = parse_fn(*m_data->query_string);
+    std::vector<std::string> processed_terms;
+    std::vector<std::uint32_t> term_ids;
+    for (auto&& term: parsed_terms) {
+        processed_terms.push_back(std::move(term.term));
+        term_ids.push_back(term.id);
+    }
+    m_data->term_ids = std::move(term_ids);
+    m_data->processed_terms = std::move(processed_terms);
+    return *this;
+}
+
+auto QueryContainer::threshold(float score) -> QueryContainer&
+{
+    m_data->threshold = score;
+    return *this;
+}
+
+auto QueryContainer::query() const -> QueryRequest
+{
+    return QueryRequest(*this);
+}
+
+template <typename T>
+[[nodiscard]] auto get(nlohmann::json const& node, std::string_view field) -> std::optional<T>
+{
+    if (auto pos = node.find(field); pos != node.end()) {
+        try {
+            return std::make_optional(pos->get<T>());
+        } catch (nlohmann::detail::exception const& err) {
+            throw std::runtime_error(fmt::format("Requested field {} is of wrong type", field));
+        }
+    }
+    return std::optional<T>{};
+}
+
+auto QueryContainer::from_json(std::string_view json_string) -> QueryContainer
+{
+    try {
+        auto json = nlohmann::json::parse(json_string);
+        QueryContainer query;
+        QueryContainerInner& data = *query.m_data;
+        bool at_least_one_required = false;
+        if (auto id = get<std::string>(json, "id"); id) {
+            data.id = std::move(id);
+        }
+        if (auto raw = get<std::string>(json, "query"); raw) {
+            data.query_string = std::move(raw);
+            at_least_one_required = true;
+        }
+        if (auto terms = get<std::vector<std::string>>(json, "terms"); terms) {
+            data.processed_terms = std::move(terms);
+            at_least_one_required = true;
+        }
+        if (auto term_ids = get<std::vector<std::uint32_t>>(json, "term_ids"); term_ids) {
```

Comment:
Is there any chance we'd have a case where the json contains raw, terms, and identifiers? And if so, which will we take as being "correct", or should we validate that they are equal somewhere?

Reply:
I was thinking of taking whatever's most processed, so say IDs, and maybe have a flag `--force-parse` or something? Or we could decide depending on whether there is lexicon provided or not. But the disadvantage of that is that someone could leave it by mistake, but not define, say, stemmer, and then it's messed up. I don't think there's a perfect solution but whatever we decide should be easy to reason about. I think it's easier when you say "always take ID first, or if no IDs, then terms, or else query": there's a clear hierarchy and if you want something else, then be explicit. Or we can say, always try recompute as much as possible, and to be explicit if you don't want to do that. I think either of these two seem the best to me at the moment.

What are your thoughts?